### PR TITLE
Beautify README

### PR DESCRIPTION
**Pull request by mhudaky**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,62 +2,62 @@
 
 DISCLAIMER: only a partial implementation of the paper called "[FastReact: In-Network Control and Caching for Industrial Control Networks using ProgrammableData Planes](https://arxiv.org/pdf/1808.06799.pdf)".
 We took a different approach, using match-action tables instead of registers.
-This implementation certainly needs some polishment.
+This implementation is only proof of concept, for the practical application it needs to be further developed (see Notes)
```
Comment (steakhal) [inline README.md]: Full stop missing at the end.

File: README.md [hunk 2]
```diff
@@ -2,62 +2,62 @@
...
+
 ## Setup
 
-I've only tested this on Windows 10 Pro.
+- only tested on Windows 10 Pro.
```
Comment (steakhal) [inline README.md]: Either remove full stop or capitalize the first letter.

File: README.md [hunk 3]
```diff
@@ -2,62 +2,62 @@
...
+5. Open two separate WSL2 shells, for h1 and h2 nodes.
 6. Make sure those shells also have the `p4app` in the `PATH` envvar.
-7. Run the `p4app exec m h1 bash` command from one and the `p4app exec m h2 bash` in the other opened shell.
+7. Run `p4app exec m h1 bash` in the first and `p4app exec m h2 bash` in the second shell.
 8. Navigate to the `/tmp` folder in both shells.
-9. Run the `python3 send.py` in one, and the `python3 receive.py` in the other shell.
-
-You will probably see some packet communication.
-The last byte of the message represents the result of the rule matching logic on the switch.
+9. Run `python3 send.py` in the first and `python3 receive.py` in the second shell, whereas the first shell ships a packet towards the second. The last byte of the packet signs if the packet satisfied the switch-defined rule.
```
Comment (steakhal) [inline README.md:26]: Excellent.

File: README.md [hunk 4]
```diff
@@ -2,62 +2,62 @@
...
 
 ## Implementation logic
 
-The user can define rules in the following form:
+The user can define rules in conjunctive normal form:
```
Comment (steakhal) [inline README.md]: disjunctive normal form

File: README.md [hunk 5]
```diff
@@ -2,62 +2,62 @@
...
 ```
 ((id00 op00 constant00)^(id01 op01 constant01)^(id02 op02 constant02)^(id03 op03 constant03))v
 ((id10 op10 constant10)^(id11 op11 constant11)^(id12 op12 constant12)^(id13 op13 constant13))v
 ((id20 op20 constant20)^(id21 op21 constant21)^(id22 op22 constant22)^(id23 op23 constant23))
 ```
-Each `id` is a sensor identifier, `op` a logical operation (except zero, which represents invalid operation) and a `constant` which we compare against.
-These (3x4x3) integer values represent a rule. As an example `4 2 7 0 0 0 0 0 0 0 0 0   0 0 0 0 0 0 0 0 0 0 0 0  0 0 0 0 0 0 0 0 0 0 0 0` represents the expression `id4 <= 7`.
-This rule is used in testing, where the rule table is populated by the `fastreact.config` file.
+where each triplet is consist of:
```
Comment (steakhal) [inline README.md]:
```diff
- where each triplet is consist of:
+ where each triplet consists of:
```

File: README.md [hunk 6]
```diff
@@ -2,62 +2,62 @@
...
+- a sensor identifier (`id`)
+- a logical operation (`op`)
+- and a constant value (`constant`)
+
+Any triplet represent a literal, where we compare the sensor value to a constant (eg. S1 > 5, where S1 is the measured value by sensor 1)"
```
Comment (steakhal) [inline README.md]:
```diff
- Any triplet represent a literal, where we compare the sensor value to a constant (eg. S1 > 5, where S1 is the measured value by sensor 1)"
+ Any triplet represent a literal, where we compare the sensor value to a constant (eg. S1 > 5, where S1 is the last measured value by sensor 1)"
```

File: README.md [hunk 7]
```diff
@@ -2,62 +2,62 @@
...
+Any rows (consist of 4 triplets) represent a condition, any of the triplets evaluating true satisfies the condtion (eg. S1 > 5 ∨ S2 = 8)
```
Comment (steakhal) [inline README.md]: It's kind of blurry to me.
A row consists of 3 groups. If any group is satisfied the rule is satisfied.
A group consists of 4 triplets. A group is satisfied only if all valid triplets (aka. the opcode is non zero) of that group are satisfied.
I think these lines require more clarification.

File: README.md [hunk 8]
```diff
@@ -2,62 +2,62 @@
...
+The whole (3x4x3) integer values represent the rule, all of the conditions (represented by rows) needs to be fulfilled to satisfy the rule.
```
Comment (steakhal) [inline README.md]: "The whole" what? row?

File: README.md [hunk 9]
```diff
@@ -2,62 +2,62 @@
...
+(triplet starting with 0 is invalid, represents empty literal and evaluates true always)
```
Comment (steakhal) [inline README.md]: Whether a triplet is evaluated or not is driven only by the opcode.

The latter is more like a knowledgable comment relating to set theory. Which says that the empty set is part of any set.
In other words, the empty requirement set is fulfilled by any inputs.
=> Rule without any valid triplet is always evaluated to true regardless of the inputs.

File: README.md [hunk 10]
```diff
@@ -2,62 +2,62 @@
...
+Take the example of `fastreact.config`, which is used for testing: `4 2 7 0 0 0 0 0 0 0 0 0   0 0 0 0 0 0 0 0 0 0 0 0  0 0 0 0 0 0 0 0 0 0 0 0`
+these (3x4x3) integers represents the expression of `id4 <= 7`.
 
 ### Packet flow
 
-Each non-sensor data packet simply forwarded to the requested destination.
-Although sensor packets handled differently.
+Each non-sensor data packet forwarded to the requested destination as default.
```
Comment (steakhal) [inline README.md]: Why not `by default`?

**Review by steakhal — CHANGES_REQUESTED** [review]
It's definitely an improvement!
I very much like the most part of this patch, though there is someplace inline which you should check more deeply.

Thank you.

File: README.md (hunk 1 above)
Comment (mhudaky) [reply README.md]: Fixed

File: README.md (hunk 2 above)
Comment (mhudaky) [reply README.md]: Fixed

File: README.md (hunk 3 above)
Comment (mhudaky) [reply README.md:26]: Thanks.

File: README.md (hunk 4 above)
Comment (mhudaky) [reply README.md]: I am quite sure it is a conjunctive normal form.

File: README.md (hunk 5 above)
Comment (mhudaky) [reply README.md]: Thanks, fixed

File: README.md (hunk 4 above)
Comment (steakhal) [reply README.md]: No it is not. [eg](https://www.google.com/imgres?imgurl=https%3A%2F%2Fimage1.slideserve.com%2F1650960%2Fdisjunctive-normal-form-2-l.jpg&imgrefurl=https%3A%2F%2Fwww.slideserve.com%2Fabbott%2Fcs1022-computer-programming-principles&docid=TWO75mpSo7Y-YM&tbnid=UEiYcPBUJ8EerM%3A&vet=1&source=sh%2Fx%2Fim)

Comment (mhudaky) [reply README.md]: Ahh, okay, now I see that we actually used disjunctive normal form in the code.
In the original paper it was conjunctive normal form.

Comment (mhudaky) [reply README.md]: Fixed

File: README.md (hunk 6 above)
Comment (mhudaky) [reply README.md]: Fixed

File: README.md (hunk 8 above)
Comment (mhudaky) [reply README.md]: I meant the 3 rows together, but I rewrote this part

File: README.md (hunk 9 above)
Comment (mhudaky) [reply README.md]: Fixed

File: README.md (hunk 10 above)
Comment (mhudaky) [reply README.md]: Fixed

File: README.md (hunk 7 above)
Comment (mhudaky) [reply README.md]: Rewrote this part to be more clear and to write about disjunctive normal form instead of conjunctive

**Review by steakhal — APPROVED** [review]
LGTM, I'm gonna merge it.

**Pull request merged**